_posts/2022-12-12-async-manual.md

```diff
@@ -740,74 +735,69 @@ public static int[][] matrixMultiplyFaster(int[][] a, int[][] b, int size) {
+- finding allocations outside the TLAB, which are done in the slow path
+- fighting with single/tens of milliseconds latency, where even the creation of heap objects matters
 
-First let's understand how new objects on a heap are created, so we have a better
-understanding of what the async-profiler shows to us.
+First, let's understand how new objects on a heap are created so we have a better
+understanding of what the async-profiler shows us.
 
 A portion of our Java heap is called an **eden**. This is a place where new objects are
-born. Let's assume for simplicity that eden is a continuous part of memory. The very efficient
+born. Let's assume for simplicity that eden is a continuous slice of memory. The very efficient
 way of allocation in such a case is called **bumping pointer**. We keep a pointer to the first 
-address of a free space:
+address of the free space:
 
 ![alt text](/assets/async-demos/alloc-1.png "alloc")
 
-When we do ```new Object()``` we simply can calculate a space needed for it and to locate 
-next free address we just need to bump the pointer by ```sizeof(Object)```:
+When we do ```new Object()```, we simply calculate its size, locate 
+the next free address and bump the pointer by ```sizeof(Object)```:
 
 ![alt text](/assets/async-demos/alloc-2.png "alloc")
 
-But there is one major problem with that technique, we have more than one thread that can 
-create new objects at the same time. We need to synchronize those calls somehow or ... we
-can give each thread a portion of eden, that is dedicated to only that thread. That portion 
-is called **TLAB** - thread local allocation buffer. In such cases each thread can use a
-**bumping pointer** at his own TLAB safely.
+But there is one major problem with that technique: We have to synchronize the object allocation if we have more than one thread that can 
+create new objects in parallel, but this is quite costly. We solve this by giving each thread a portion of eden dedicated to only that thread. This portion 
+is called **TLAB** - thread-local allocation buffer. With this, each thread can use **bumping pointer** at its TLAB safely and in parallel.
 
-Introducing TLABs creates two more issues that JVM needs to deal:
+Introducing TLABs creates two more issues that the JVM needs to deal with:
 
-- a thread can allocate an object, but there is not enough space in its TLAB - in that case
-  if there is still a space in eden JVM creates new TLAB for that thread
-- a thread can allocate a big object, so it's not optimal to use TLAB mechanism - in that case
-  JVM will use _slow path_ of allocation that allocates the object directly in eden
+- a thread can allocate an object, but there is not enough space in its TLAB - the JVM creates a new TLAB if there is still space in eden
+- a thread can allocate a big object, so it's not optimal to use the TLAB mechanism - the JVM will use the _slow path_ of the allocation that allocates the object directly in eden
```

Comment:
G1 Humongous objects are created directly in the old-gen, forgot about it before.